applications/ecallmgr/src/ecallmgr_fs_channel.erl

```diff
@@ -387,23 +431,69 @@ code_change(_OldVsn, State, _Extra) ->
                                              ,{<<"uuid">>, UUID}
                                             ])
     end.
 
+-spec handle_channel_req(ne_binary(), ne_binary(), wh_proplist(), atom(), pid()) -> 'ok'.
+handle_channel_req(UUID, FetchId, Props, Node, Pid) ->
+    wh_amqp_channel:consumer_pid(Pid),
+    ForUUID = props:get_value(<<"refer-for-channel-id">>, Props),
+    {'ok', ForChannel} = fetch(ForUUID, 'proplist'),
+    case fetch_channel(UUID) of
+        'undefined' -> channel_not_found(Node, FetchId);
+        Channel ->
+            [Uri] = nksip_parse_uri:uris(props:get_value(<<"switch_url">>, Channel)),
+            URL = nksip_unparse:ruri(
+                    #uri{user= props:get_value(<<"refer-to-user">>, Props)
+                         ,domain= props:get_value(<<"realm">>, Channel)
+                         ,opts=[{<<"fs_path">>, nksip_unparse:ruri(Uri#uri{user= <<>>})}]
+                        }),
+            build_channel_resp(FetchId, Props, Node, URL, ForChannel, channel_ccvs(Channel))
+    end.
+
+-spec build_channel_resp(ne_binary(), wh_proplist(), atom(), ne_binary(), wh_proplist(), wh_proplist()) -> 'ok'.
+build_channel_resp(FetchId, Props, Node, URL, Channel, ChannelVars) ->
+%% NOTE
+%% valid properties to return are
+%% sip-url , dial-prefix, absolute-dial-string, sip-profile (defaulted to current channel profile)
+%% freeswitch formats the dial string with the following logic
+%% if absolute-dial-string => %s%s [dial-prefix, absolute-dial-string]
+%% else => %ssofia/%s/%s [dial-prefix, sip-profile, sip-url]
+    Resp = props:filter_undefined(
+             [{<<"sip-url">>, URL}
+              ,{<<"dial-prefix">>, channel_resp_dialprefix(Props, Channel, ChannelVars)}
+             ]),
+    try_channel_resp(FetchId, Node, Resp).
+
+-spec channel_resp_dialprefix(wh_proplist(), wh_proplist(), wh_proplist()) -> ne_binary().
+channel_resp_dialprefix(ReqProps, _Channel, ChannelVars) ->
+    Props = props:filter_undefined(
+              [{<<"sip_invite_domain">>, props:get_value(<<"Realm">>, ChannelVars)}
+               ,{<<"sip_h_X-Core-UUID">>, props:get_value(<<"Core-UUID">>, ReqProps)}
+               ,{<<"sip_h_X-ecallmgr_replaces-call-id">>, props:get_value(<<"replaces-call-id">>, ReqProps)}
+               ,{<<"sip_h_X-ecallmgr_refer-from-channel-id">>, props:get_value(<<"refer-from-channel-id">>, ReqProps)}
+               ,{<<"sip_h_X-ecallmgr_refer-for-channel-id">>, props:get_value(<<"refer-for-channel-id">>, ReqProps)}
+               ,{<<"sip_h_X-ecallmgr_Account-ID">>, props:get_value(<<"Account-ID">>, ChannelVars)}
+               ,{<<"sip_h_X-ecallmgr_Realm">>, props:get_value(<<"Realm">>, ChannelVars)}
+              ]),
+    X = string:join(
+          [wh_util:to_list(<<K/binary, "='", (wh_util:to_binary(V))/binary, "'">>)
+           || {K, V}  <- Props
+          ]
+          ,","
+         ),
+    wh_util:to_binary(["[", X, "]"]).
```

Comment:
I'm confused by all these conversionnnnns
```erlang
fs_props_to_string([]) -> <<"[]">>; %% dunno if useful
fs_props_to_string([{Hk,Hv}|T]) ->
    Rest = << <<",", K/binary, "='", (wh_util:to_list(V))/binary, "'">> || {K,V} <- T >>,
    <<"[", Hk/binary, "='", (wh_util:to_binary(Hv))/binary, "'", Rest/binary, "]">>.
```

…much more readable, isn't it? :d